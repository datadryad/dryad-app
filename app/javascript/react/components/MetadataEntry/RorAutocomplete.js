import React, {useState} from 'react';
import {useCombobox} from 'downshift';
import {comboboxStyles, items, menuStyles} from './shared';
import axios from "axios";

export default function RorAutocomplete() {
  const [inputItems, setInputItems] = useState(items);
  const {
    isOpen,
    getToggleButtonProps,
    getLabelProps,
    getMenuProps,
    getInputProps,
    getComboboxProps,
    highlightedIndex,
    getItemProps,
  } = useCombobox({
    items: inputItems,
    onInputValueChange: ({inputValue}) => {
      const result = [];
      // URL = https://api.ror.org/organizations
      // query="name"
      axios.get('https://api.ror.org/organizations', { params: {query: "University of California"},
            headers: {'Content-Type': 'application/json; charset=utf-8', Accept: 'application/json'} } )
          .then((data) => {
            if (data.status !== 200) {
              console.log('Response failure not a 200 response');
            }else{
              // const myList = data.data.items.map((item) => item.name);
              const myList = data.data.items;
              console.log(myList);
              setInputItems(myList,)
            }
          });
      /*setInputItems(
          items.filter((item) =>
              item.toLowerCase().startsWith(inputValue.toLowerCase()),
          ),
      )*/
    },
    itemToString: (item) => item?.name || '',
  });

  return (
      <div>
        <label {...getLabelProps()}>Choose an element:</label>
        <div style={comboboxStyles} {...getComboboxProps()}>
          <input className='c-input__text' {...getInputProps()} />
          <button
              type="button"
              {...getToggleButtonProps()}
              aria-label="toggle menu"
          >
            &#8595;
          </button>
        </div>
        <ul {...getMenuProps()} style={menuStyles}>
          {isOpen &&
          inputItems.map((item, index) => (
              <li
                  {...getItemProps({item, index})}
                  style={
                    highlightedIndex === index ? {backgroundColor: '#bde4ff'} : {}
                  }
                  key={item.id}
              >
                {item.name}
              </li>
          ))}
        </ul>
      </div>
  );
};
