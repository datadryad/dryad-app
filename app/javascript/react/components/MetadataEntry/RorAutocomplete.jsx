import React, {useEffect, useRef, useState} from 'react';
import axios from 'axios';
import PropTypes from 'prop-types';
import GenericNameIdAutocomplete from './GenericNameIdAutocomplete';

export default function RorAutocomplete({
  formRef, acText, setAcText, acID, setAcID, controlOptions,
}) {
  // control options: htmlId, labelText, isRequired (t/f)

  // in order to use this component, we need to track the state of the autocomplete text and the autocomplete id
  // https://www.freecodecamp.org/news/what-is-lifting-state-up-in-react/ is a better functional example than the react docs.
  // also tracking "autoBlurred" since we need to know when things exit to trigger form resubmission or sending to server.
  const prevText = useRef(acText);
  const prevID = useRef(acID);
  const [autoBlurred, setAutoBlurred] = useState(false);

  // do something when blurring from the autocomplete, passed up here, probably want to save on blur, but save
  // action may be different depending on autocomplete context inside another form or may save directly.
  useEffect(() => {
    if (autoBlurred) {
      if (!acText) {
        setAcID('');
      }
      /* it seems like the current way for this to work within authors is to add elements named
            "author[affiliation][long_name]" and "author[affiliation][ror_id]" that have correct values and resubmit
            the form.
           */
      if (prevText.current !== acText || prevID.current !== acID) {
        // from the ref, submit the Formik form above me
        formRef.current.handleSubmit();
      }
      prevText.current = acText;
      prevID.current = acID;
      setAutoBlurred(false);
    }
  }, [autoBlurred]);

  /* supplyLookupList returns a promise that will supply a list of js objects that
     can have their properties used to create the autocomplete list.
     It is required to be passed in so we can get lists from various data sources which may vary across different
     autocompletes for a generic case.
   */
  const supplyLookupList = (qt) => axios.get('/stash_datacite/affiliations/autocomplete', {
    params: {query: qt},
    headers: {'Content-Type': 'application/json; charset=utf-8', Accept: 'application/json'},
  })
    .then((data) => {
      if (data.status !== 200) {
        return [];
        // raise an error here if we want to catch it and display something to user or do something else
      }
      return data.data.map((i) => {
        i.display = (<>{i.name}<br /><small>{i.acronyms.length > 1 && `(${i.acronyms.join(', ')}) `}{i.country}</small></>);
        return i;
      });
    });

  // Given a js object from list (supplyLookupList above) it returns the string name
  const nameFunc = (item) => (item?.name || '');

  // Given a js object from list (supplyLookupList above) it returns the unique identifier
  const idFunc = (item) => item.id;

  return (
    <GenericNameIdAutocomplete
      acText={acText || ''}
      setAcText={setAcText}
      acID={acID}
      setAcID={setAcID}
      setAutoBlurred={setAutoBlurred}
      supplyLookupList={supplyLookupList}
      nameFunc={nameFunc}
      idFunc={idFunc}
      controlOptions={controlOptions}
    />
  );
}

RorAutocomplete.propTypes = {
  formRef: PropTypes.object.isRequired,
  acText: PropTypes.string.isRequired,
  setAcText: PropTypes.func.isRequired,
  acID: PropTypes.string.isRequired,
  setAcID: PropTypes.func.isRequired,
  controlOptions: PropTypes.object.isRequired,
};
