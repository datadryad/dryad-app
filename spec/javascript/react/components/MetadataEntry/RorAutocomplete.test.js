/**
 * @jest-environment jsdom
 */

import {shallow} from "enzyme"; // mount vs shallow (for superficial test)
import toJSON from "enzyme-to-json";
import React from "react";
import RorAutocomplete from "../../../../../app/javascript/react/components/MetadataEntry/RorAutocomplete.js";

describe('RorAutocomplete', () => {
  it("renders a basic autocomplete form", () => {

    const info = {name: 'Institute for Fermentation', id: 'https://ror.org/05nq89q24',
      'controlOptions': { htmlId: "instit_affil_1234", labelText: 'Institutional Affiliation', isRequired: true } }

    const wrapper = shallow(
        <RorAutocomplete {...info} />
    );

    expect(toJSON(wrapper)).toMatchSnapshot(); // this matches snapshot unless it's updated
  });

  it('should return list of institutions in dropdown', () => {
  });

  it('should select institution from dropdown and fill hidden fields', () => {
  });
});