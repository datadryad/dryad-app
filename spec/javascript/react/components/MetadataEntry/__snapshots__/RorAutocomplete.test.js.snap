// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RorAutocomplete renders a basic autocomplete form 1`] = `
<Fragment>
  <GenericNameIdAutocomplete
    acID="https://ror.org/05nq89q24"
    acText="Institute for Fermentation"
    controlOptions={
      Object {
        "htmlId": "instit_affil_1234",
        "isRequired": true,
        "labelText": "Institutional Affiliation",
      }
    }
    idFunc={[Function]}
    nameFunc={[Function]}
    setAcID={[Function]}
    setAcText={[Function]}
    setAutoBlurred={[Function]}
    supplyLookupList={[Function]}
  />
  <input
    className="js-affil-longname"
    name="author[affiliation][long_name]"
    type="hidden"
    value="Institute for Fermentation"
  />
  <input
    className="js-affil-id"
    name="author[affiliation][ror_id]"
    type="hidden"
    value="https://ror.org/05nq89q24"
  />
</Fragment>
`;
